### PR TITLE
Add new parameterization of snow unloading

### PR DESCRIPTION
**Pull request by arbennett**
This work with @cassielumbrazo adds a new decision along with a new parameterization. The decision is `snowUnload`, which will have 2 options. `meltDripUnload` was the previously default implementation. The new parameterization, `windUnload` implements a snow unloading scheme described in [A new snow cover fraction parametrization for the ECHAM4 GCM](https://link.springer.com/article/10.1007%2Fs003820100153) eq 13. 

Further testing is needed before this should be formally reviewed, but appears to be working as intended:
image

This feature adds 4 new parameters which need to be specified. They are, along with some default values from Roesch et al:

```
! Minimum temperature for temperature unloading (T< minTempUnloading causes no unloading) (K)
minTempUnloading          |       270.16 |       260.16 |       273.16
! Minimum wind for wind unloading (v < minWindUnloading causes no unloading) (m/s)
minWindUnloading          |       0.0000 |       0.0000 |       10.000
! Scales the speed at which unloading due to temperature occurs (inversely proportional) (K*s)
rateTempUnloading         |      1.87d+5 |       1.0d+5 |       3.0d+5
! Scales the speed at which unloading due to wind occurs (inversely proportional) (m)
rateWindUnloading         |      1.56d+5 |       1.0d+5 |       3.0d+5
```

### CODE REVIEW

**Comment by bartnijssen** [comment]
@arbennett : `Further testing is needed before this should be formally reviewed`

Do you mean that further testing is needed before code review or before this can be reviewed science-wise? Just want to make sure I understand whether you want me to do anything at this time.

**Comment by arbennett** [comment]
@bartnijssen - Further testing is needed before review of the code. You don't have to do anything yet.

**Comment by arbennett** [comment]
It appears this is okay to be reviewed now.

File: build/source/engine/canopySnow.f90 [hunk 1]
```diff
@@ -123,10 +132,14 @@ subroutine canopySnow(&
 
  ! model prognostic variables (input/output)
  scalarCanopyIce           => prog_data%var(iLookPROG%scalarCanopyIce)%dat(1),             & ! intent(inout): [dp] mass of ice on the vegetation canopy (kg m-2)
+ scalarCanairTemp          => prog_data%var(iLookPROG%scalarCanairTemp)%dat(1),             & ! intent(input): [dp] temperature of the canopy air space (k)
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: align `&`

File: build/source/engine/canopySnow.f90 [hunk 2]
```diff
@@ -96,10 +98,16 @@ subroutine canopySnow(&
...
  real(dp)                      :: scalarCanopyIceIter        ! trial value for mass of ice on the vegetation canopy (kg m-2) (kg m-2)
+ real(dp)                      :: accum
  real(dp)                      :: flux                       ! net flux (kg m-2 s-1)
  real(dp)                      :: delS                       ! change in storage (kg m-2)
  real(dp)                      :: resMass                    ! residual in mass equation (kg m-2)
+ real(dp)                      :: tempUnloadingFun           ! temperature unloading functions, Eq. 14 in Roesch et al. 2001
+ real(dp)                      :: windUnloadingFun           ! temperature unloading functions, Eq. 15 in Roesch et al. 2001
  real(dp),parameter            :: convTolerMass=0.0001_dp    ! convergence tolerance for mass (kg m-2)
+ real(dp),parameter            :: C_1=-270.15_dp             ! constant 1 for wind unloading (K)
+ real(dp),parameter            :: C_2=1.87d+5           ! constant 2 for wind unloading (K s)
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: align comment

File: build/source/engine/canopySnow.f90 [hunk 3]
```diff
@@ -96,10 +98,16 @@ subroutine canopySnow(&
...
+ real(dp),parameter            :: C_3=1.56d+5           ! constant 3 for wind unloading (m)
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: align comment

File: build/source/engine/canopySnow.f90 [hunk 4]
```diff
@@ -96,10 +98,16 @@ subroutine canopySnow(&
  real(dp)                      :: throughfallDeriv           ! derivative in throughfall flux w.r.t. canopy storage (s-1)
  real(dp)                      :: unloadingDeriv             ! derivative in unloading flux w.r.t. canopy storage (s-1)
  real(dp)                      :: scalarCanopyIceIter        ! trial value for mass of ice on the vegetation canopy (kg m-2) (kg m-2)
+ real(dp)                      :: accum
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: add comment that explains what `accum` is. Also consider using a more informative variable name (and carry it throughout)

File: build/source/engine/canopySnow.f90 (hunk 3 above)
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: Consider using more informative constant names than `C_1`, `C_2`, `C_3`

File: build/source/engine/canopySnow.f90 (hunk 1 above)
Comment (bartnijssen) [reply build/source/engine/canopySnow.f90]: Use `intent(in)` rather than `intent(input)`

File: build/source/engine/canopySnow.f90 [hunk 5]
```diff
@@ -123,10 +132,14 @@ subroutine canopySnow(&
 
  ! model prognostic variables (input/output)
  scalarCanopyIce           => prog_data%var(iLookPROG%scalarCanopyIce)%dat(1),             & ! intent(inout): [dp] mass of ice on the vegetation canopy (kg m-2)
+ scalarCanairTemp          => prog_data%var(iLookPROG%scalarCanairTemp)%dat(1),             & ! intent(input): [dp] temperature of the canopy air space (k)
 
  ! model fluxes (input)
  scalarSnowfall            => flux_data%var(iLookFLUX%scalarSnowfall)%dat(1),              & ! intent(in): [dp] computed snowfall rate (kg m-2 s-1)
  scalarCanopyLiqDrainage   => flux_data%var(iLookFLUX%scalarCanopyLiqDrainage)%dat(1),     & ! intent(in): [dp] liquid drainage from the vegetation canopy (kg m-2 s-1)
+ scalarWindspdCanopyTop    => flux_data%var(iLookFLUX%scalarWindspdCanopyTop)%dat(1),      & ! intent(in): [dp] windspeed at the top of the canopy (m s-1)
+ scalarWindspdCanopyBottom => flux_data%var(iLookFlux%scalarWindspdCanopyBottom)%dat(1),   & ! intent(in): [dp] windspeed at the height of the bottom of the canopy (m s-1)
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: drop the `height of the` to be consistent with the previous comment (previous line)

File: build/source/engine/canopySnow.f90 [hunk 6]
```diff
@@ -158,86 +170,66 @@ subroutine canopySnow(&
...
+         scalarCanopySnowUnloading = snowUnloadingCoeff*scalarCanopyIceIter
+         unloadingDeriv            = snowUnloadingCoeff
+     else if (ixSnowUnload==windUnload) then
+         tempUnloadingFun = (C_1 + scalarCanairTemp) / C_2   ! (s-1)
+         windUnloadingFun = abs(scalarWindspdCanopyTop) / C_3     ! (s-1)
+         ! No snow unloading if T < -3C
+         if (tempUnloadingFun < 0) then
+             tempUnloadingFun = 0._dp
+         end if
+         ! define the "windySnow"  Roesch et al. 2001 parameterization, Eq. 13 in Roesch et al. 2001
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: Are you `defining` it here or simply `using` or `implementing it?

File: build/source/engine/canopySnow.f90 [hunk 7]
```diff
@@ -158,86 +170,66 @@ subroutine canopySnow(&
...
-     elseif(airtemp_degC > -3._dp)then; leafScaleFactor = 1.5_dp*airtemp_degC + 5.5_dp
-                                  else; leafScaleFactor = 1.0_dp
+     ! ** compute unloading
+     if (ixSnowUnload==meltDripUnload) then
+         scalarCanopySnowUnloading = snowUnloadingCoeff*scalarCanopyIceIter
+         unloadingDeriv            = snowUnloadingCoeff
+     else if (ixSnowUnload==windUnload) then
+         tempUnloadingFun = (C_1 + scalarCanairTemp) / C_2   ! (s-1)
+         windUnloadingFun = abs(scalarWindspdCanopyTop) / C_3     ! (s-1)
+         ! No snow unloading if T < -3C
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: Comment does not appear to match the line that follows and is confusing `<-3` vs `<0` in the following line. Should this be a hard-coded constant or something that should be set in a parameter file?

File: build/source/engine/canopySnow.f90 [hunk 8]
```diff
@@ -246,6 +238,7 @@ subroutine canopySnow(&
  ! *****
  ! update mass of ice on the canopy (kg m-2)
  scalarCanopyIce = scalarCanopyIceIter
+ !write(*,*) scalarCanopySnowUnloading !, scalarThroughfallSnow, flux, delS
```
Comment (bartnijssen) [inline build/source/engine/canopySnow.f90]: remove debug statements - same for the following lines

File: build/source/engine/mDecisions.f90 [hunk 9]
```diff
@@ -523,7 +526,7 @@ subroutine mDecisions(err,message)
  select case(trim(model_decisions(iLookDECISIONS%snowIncept)%cDecision))
   case('stickySnow'); model_decisions(iLookDECISIONS%snowIncept)%iDecision = stickySnow        ! maximum interception capacity an increasing function of temerature
   case('lightSnow' ); model_decisions(iLookDECISIONS%snowIncept)%iDecision = lightSnow         ! maximum interception capacity an inverse function of new snow density
-  case default
+   case default
```
Comment (bartnijssen) [inline build/source/engine/mDecisions.f90]: align `case`

File: build/source/engine/time_utils.f90 [hunk 10]
```diff
@@ -98,6 +98,7 @@ subroutine extractTime(refdate,iyyy,im,id,ih,imin,dsec,ih_tz,imin_tz,dsec_tz,err
 
  ! FIELD 2: Advance to the ih:imin:dsec string
  istart=nsub+1
+ if(istart>n) return  ! no more time info
```
Comment (bartnijssen) [inline build/source/engine/time_utils.f90]: What is the purpose of this? Seems like this should be a separate PR

File: build/source/engine/mDecisions.f90 [hunk 11]
```diff
@@ -625,6 +628,15 @@ subroutine mDecisions(err,message)
    err=10; message=trim(message)//"unknown option for new snow density [option="//trim(model_decisions(iLookDECISIONS%snowDenNew)%cDecision)//"]"; return
  end select
 
+ ! choice of snow unloading from canopy
+ select case(trim(model_decisions(iLookDECISIONS%snowUnload)%cDecision))
+  case('meltDripUnload','notPopulatedYet'); model_decisions(iLookDECISIONS%snowUnload)%iDecision = meltDripUnload  ! Hedstrom and Pomeroy (1998), Storck et al 2002 (snowUnloadingCoeff & ratioDrip2Unloading)
+  case('windUnload');                   model_decisions(iLookDECISIONS%snowUnload)%iDecision = windUnload          ! Roesch et al 2001, formulate unloading based on wind and temperature
```
Comment (bartnijssen) [inline build/source/engine/mDecisions.f90]: alignment confusing

File: build/source/engine/time_utils.f90 (hunk 10 above)
Comment (arbennett) [reply build/source/engine/time_utils.f90]: Hm, this seems to be vestigial. It has been removed.

File: build/source/engine/mDecisions.f90 (hunk 9 above)
Comment (arbennett) [reply build/source/engine/mDecisions.f90]: Done

File: build/source/engine/canopySnow.f90 (hunk 8 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: Done.

File: build/source/engine/canopySnow.f90 (hunk 7 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: This has to do with the `C_1 + scalarCanairTemp` portion in `tempUnloadingFun`. That part evaluates to `scalarCanairTemp - 270.15`, hence the 3C. It seems like a better approach may be to be set in the parameter file, along with `C_2` and `C_3`. Along with better names, of course.

File: build/source/engine/canopySnow.f90 (hunk 4 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: Removed this, it was not used anywhere else.

File: build/source/engine/canopySnow.f90 (hunk 3 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: Done

File: build/source/engine/canopySnow.f90 (hunk 2 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: Done

File: build/source/engine/canopySnow.f90 (hunk 1 above)
Comment (arbennett) [reply build/source/engine/canopySnow.f90]: Done, and moved to other `intent(in)` section

File: build/source/engine/mDecisions.f90 (hunk 11 above)
Comment (arbennett) [reply build/source/engine/mDecisions.f90]: I updated the alignment.

**Comment by cassielumbrazo** [comment]
I have evaluated the algorithm and tested all new parameters at multiple sites. It appears to be working in all tests. An example at Niwot Ridge, CO below. 
image

Another example, testing the minimum wind speed threshold in meters/sec. 
image

I am also updating documentation describing this new decision.

**Comment by bartnijssen** [comment]
Will be merged before SUMMA V3.0 after completing series of PRs related to changes to decision file, file manager, and some general cleanup
